ratpack-retrofit2/src/main/java/ratpack/retrofit/RatpackRetrofit.java

```diff
@@ -0,0 +1,169 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ratpack.retrofit;
+
+import com.google.common.base.Preconditions;
+import ratpack.func.Action;
+import ratpack.http.client.HttpClient;
+import ratpack.retrofit.internal.RatpackCallAdapterFactory;
+import ratpack.retrofit.internal.RatpackCallFactory;
+import ratpack.util.Exceptions;
+import retrofit2.Converter;
+import retrofit2.Retrofit;
+import retrofit2.converter.scalars.ScalarsConverterFactory;
+
+import java.net.URI;
+
+/**
+ * Builder for providing integration of Retrofit2 with Ratpack's {@link HttpClient}.
+ * <p>
+ * This class allows for creating declarative type-safe interfaces that represent remote HTTP APIs.
+ * Using this adapter allows for defining the interfaces to return {@link ratpack.exec.Promise} types which will be fulfilled by Ratpack's http client.
+ *
+ * <pre class="java">{@code
+ * import ratpack.exec.Promise;
+ * import ratpack.http.client.HttpClient;
+ * import ratpack.retrofit.RatpackRetrofit;
+ * import ratpack.server.PublicAddress;
+ * import ratpack.test.embed.EmbeddedApp;
+ * import retrofit2.http.GET;
+ *
+ * import static org.junit.Assert.*;
+ *
+ * public class ExampleRetrofitClient {
+ *
+ *   public interface HelloService {
+ *
+ *     {@literal @}GET("hello") Promise<String> hello();
+ *   }
+ *
+ *   public static void main(String... args) throws Exception {
+ *
+ *     EmbeddedApp.fromHandlers(chain -> {
+ *         chain.get(ctx -> {
+ *           PublicAddress address = ctx.get(PublicAddress.class);
+ *           HttpClient httpClient = ctx.get(HttpClient.class);
+ *           HelloService service = RatpackRetrofit.builder(httpClient)
+ *             .uri(address.get())
+ *             .build(HelloService.class);
+ *
+ *           ctx.render(service.hello());
+ *         });
+ *         chain.get("hello", ctx -> ctx.render("hello"));
+ *       }
+ *     ).test(testHttpClient -> {
+ *       assertEquals("hello", testHttpClient.getText());
+ *     });
+ *   }
+ * }
+ * }</pre>
+ *
+ */
+public class RatpackRetrofit {
+
+  /**
+   * Create a new builder for creating Retrofit implementations.
+   *
+   * @param httpClient the http client for Retrofit to use to make requests.
+   *
+   * @return a builder
+   */
+  public static Builder builder(HttpClient httpClient) {
+    return new Builder(httpClient);
+  }
+
+  public static class Builder {
+
+    private HttpClient httpClient;
+    private Action<? super Retrofit.Builder> builderAction = Action.noop();
+    private URI uri;
+
+    public Builder(HttpClient httpClient) {
+      this.httpClient = httpClient;
```

Comment:
I think maybe I can do away with this by using `Execution.current().get(HttpClient.class)` down when the request is actually made?